### DISCUSSION TITLE
Marp CLI with nextjs

### DISCUSSION THREAD

**Discussion by ruthkangyr** [marp-team/marp#499] [Q&A] [answered]
Hi,

Im trying to use the marpCli in my nextjs webpage to obtain the output (pdf or pptx). I did this but it does not work.

```
const tempMarkdownPath = join(tmpdir(), ${uuidv4()}.md)
await writeFile(tempMarkdownPath, markdownContent)

 
const tempPdfPath = join(tmpdir(), ${uuidv4()}.pdf)
const output = await marpCli([tempMarkdownPath, '--pdf', '--output', tempPdfPath])
```

Can i ask if there are some working examples of this implementation to use, or do is there any suggestion on this case? Thank you

**Comment by yhatt** [comment, marked as answer]
If you are attempting to run Marp CLI on the server side with Next.js App Router, adding `@marp-team/marp-cli` pacakge to the [`serverComponentsExternalPackages`](https://nextjs.org/docs/app/api-reference/next-config-js/serverComponentsExternalPackages) setting in `next.config.js` may make it work correctly.

```javascript
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    serverComponentsExternalPackages: ['@marp-team/marp-cli'],
  },
}

export default nextConfig
```

**Reply by ruthkangyr** [reply]
it works! Thank you so much. 

I would also like to ask about columns. Below is my code, and when I use Marp CLI to download the pdf/pptx, the columns does not show, but I am able to view it in vscode. Is there any changes I have to make? I believe that there may be settings that I did not set correctly but I can't seem to figure it out. Thank you!

```
---
marp: true
backgroundColor: white
style: |

  .columns-left {
      background: yellow;
      padding: 20px;
  }

  .columns-right {
      background: beige;
      padding: 20px;
  }

---

# Multi columns in Marp slide

<div class="columns">
<div class="columns-left">

## Column 1

Lorem ipsum dolor sit amet consectetur adipisicing elit. Voluptas eveniet, corporis commodi vitae accusamus obcaecati dolor corrupti eaque id numquam officia velit sapiente incidunt dolores provident laboriosam praesentium nobis culpa.

</div>
<div class="columns-right">

## Column 2

Tempore ad exercitationem necessitatibus nulla, optio distinctio illo non similique? Laborum dolor odio, ipsam incidunt corrupti quia nemo quo exercitationem adipisci quidem nesciunt deserunt repellendus inventore deleniti reprehenderit at earum.

</div>
</div>
```

On vscode:
<img width="384" alt="image" src="https://github.com/marp-team/marp/assets/77619761/02cc4eed-2341-4353-9ddf-e2de5189ae16">

Downloaded pdf:
<img width="605" alt="image" src="https://github.com/marp-team/marp/assets/77619761/8e411074-e97b-47c4-9bc5-2c87c67d12fc">

**Reply by yhatt** [reply]
For security reasons, Marp tools do not allow the use of HTML tags within Markdown by default. To enable HTML tags in Marp CLI, add the [`--html`](https://github.com/marp-team/marp-cli#:~:text=boolean%20%7C%20object-,%2D%2Dhtml,-Enable%20or%20disable) option.

> [!IMPORTANT]
> If you are attempting to use Marp CLI on a web server, especially if that server is publicly accessible and allows passing unpredictable Markdown contents, _you should pay attention particularly about allowing HTML tags._
>
> Something malicious may try to render Marp Markdown with injecting insecure HTML tags, and **may allow to attack the server, consume resources of the server, read the important information inside the server.**
>
> If you have such requirements, consider [passing an allowlist to the `html` option](https://github.com/marp-team/marp-core?tab=readme-ov-file#html-boolean--object) through a configuration JSON file (`--config-file`). Only permitted tags and attributes can be rendered.

```javascript
import { marpCli, defineConfig } from '@marp-team/marp-cli'
import path from 'node:path'
import os from 'node:os'
import { randomUUID } from 'node:crypto'
import { writeFile } from 'node:fs/promises'

const tmpPdfPath = path.join(os.tmpdir(), `${randomUUID()}.pdf`)
const tmpConfPath = path.join(os.tmpdir(), `${randomUUID()}.json`);

await writeFile(
  tmpConfPath,
  JSON.stringify(
    defineConfig({
      pdf: true,
      output: tmpPdfPath,

      // Allow HTML only `<div>` tags and its `class` attribute
      html: {
        div: ['class'],
      },
    })
  )
);

await marpCli([tmpMarkdownPath, '-c', tmpConfPath]);
```

**Reply by ruthkangyr** [reply]
it works! thank you so much

**Reply by ruthkangyr** [reply]
Hello, I am deploying my web application to vercel but I think there will be issues with CHROME_PATH etc. 

```
The CHROME_PATH environment variable must be set to a Chrome/Chromium executable no older than Chrome stable.

Error during conversion: Failed converting Markdown. (You have to install Google Chrome, Chromium, or Microsoft Edge to convert slide deck with current options.)
```

I just want to check if you happen to know of any solutions that will work for vercel with MarpCLI, instead of just manually setting a CHROME_PATH since it will differ between users. I have looked through the different threads and a lot is targetted at WSL/WSL2. 

Thank you!

**Reply by yhatt** [reply]
[Marp CLI example](https://github.com/yhatt/marp-cli-example/blob/7b3e5dffcff0e84b939cad95339501736a2e135e/package.json#L12) has a solution for getting work Marp CLI on Vercel's build time. Install Chrome and set `CHROME_PATH` env as its path before building.

For server-side use, you may have to set appropriate `CHROME_PATH` before calling `marpCli()`. Example of using [`chrome-aws-lambda`](https://github.com/alixaxel/chrome-aws-lambda):

```javascript
import chromium from 'chrome-aws-lambda'

// Store the original CHROME_PATH environment variable
const { CHROME_PATH } = process.env

try {
  // Set the CHROME_PATH environment variable to the path of the Chromium binary
  process.env.CHROME_PATH = await chromium.executablePath
  
  await marpCli(/* ... */)
} finally {
  // Restore the original CHROME_PATH environment variable
  process.env.CHROME_PATH = CHROME_PATH
}
```

**Reply by ruthkangyr** [reply]
sorry just to check, is this the line I have to add to package.json for build? or if I'm using server side then I won't need to do this? because just setting `CHROME_PATH` before calling `marpCli()` doesn't work

`MARP_USER=root CHROME_PATH=$(npx -y @puppeteer/browsers@latest install chrome@stable --path $(realpath ./tmp) | awk '{print $2}')`

**Reply by yhatt** [reply]
It's not possible to provide a clear answer because I could not know how your Next.js app uses Marp CLI specifically. Could you share a [minimal reproducible Next.js project?](https://en.wikipedia.org/wiki/Minimal_reproducible_example)

For instance, if Marp CLI is being used for rendering pages by App Router, the Chrome executable binary must be findable at the time of the initial app build. If the `CHROME_PATH` env is not set during build, the build and deployment will fail. (https://vercel.com/docs/projects/environment-variables)

So this setting should not be necessary if your application has already been built and deployed. 

In cases where Marp CLI is run on the server side, such as with API routes, it is necessary to set the browser path with `CHROME_PATH` env at server runtime.

> [!NOTE]
> It's well known that running a headless Chrome browser on the Vercel infrastructure is difficult due to a size limit of 50 MB for the build result.
> https://gist.github.com/kettanaito/56861aff96e6debc575d522dd03e5725
>
> The `chrome-aws-lambda` package, mentioned at the example provided before, has not been maintained for some time, and may cause errors due to incompatibility with the Chrome version required by the latest Marp CLI. [`@sparticuz/chromium`](https://github.com/Sparticuz/chromium) (or `@sparticuz/chromium-min` to keep the build size limit) is a good alternative.

**Reply by ruthkangyr** [reply]
thank you so much!! I'm using MarpCLI on server side and I got it to work by setting the `CHROME_PATH` on build as suggested and I think like you mentioned, the issue was with the `chrome-aws-lambda` package, after switching to `@sparticuz/chromium-min` it worked!!
